### ISSUE TITLE
negative RGB values when mixing

### ISSUE THREAD

**Issue by bnehring38** [waacton/Unicolour#3]
I am using this library to test gradient blending approaches. I must be doing something wrong, because for a particular gradient I am getting negative RGB values.

Specifically, I am blending an Rgb255 Unicolour of (0,0,84) with an Rgb255 of (0,84,250). When a do a mix in the Oklab ColourSpace, the R value goes slightly negative. And actually, my next mix of (0,84,250) with (0,146,4) quickly gets negative - at an _amount_ of 0.025 I get a result of (-9,88,245), and the red gets down to -46 later.

Is the premultiplied alpha affecting this? Other than that this seems like a straightforward interpolation, where nothing should get negative.

**Comment by waacton** [comment]
@bnehring38 you're not doing anything wrong, more likely that interpolating through non-RGB spaces just isn't always intuitive.

The colours `Rgb255 (0, 0, 84)` and `Rgb255 (0, 84, 250)` are right on the boundary of `Oklab` space. To get from one to the other in `Oklab` space, you have to pass through colours that can't be displayed by `sRGB` (colours that are outwith the 0 - 255 range). In this case, the colour proportions require so little "sRGB red" that the only way to achieve that is with negative red.

I made a [3D colour space visualisation tool](https://play.unity.com/en/games/6826f61f-3806-4155-b824-7866b1edaed7/3d-colour-space-visualisation-unicolour-demo) that was just a proof-of-concept, but might help clarify what's going on. This is an exaggerated example but your interpolation is doing something like this red arrow, traversing a region that `sRGB` can't handle.

![image](https://github.com/user-attachments/assets/27d34d6c-e78e-4fd8-85e1-6f360ae8f1da)

You can approximate what these non-displayable interpolated colours look like by mapping them into the RGB gamut, the two easiest ways in Unicolour currently being:

1. Clipping RGB values (quick and easy, can use `unicolour.Rgb.Byte255.ConstrainedTriplet` for convenience, and is the approach I use [in my gradient examples](https://github.com/waacton/Unicolour#gradients))
2. `Oklch`-based mapping from the [CSS specification](https://www.w3.org/TR/css-color-4/#gamut-mapping) (potentially slow but aims to retain the colour's hue, using [`unicolour.MapToGamut()`](https://github.com/waacton/Unicolour#map-colour-into-display-gamut))

Hope that helps clear things up.

**Comment by bnehring38** [comment]
Thank you for the detailed comment. I see what is going on now, this was very helpful!

I'll try the ConstrainedTriplet helper.

Is there any documentation on the available functions? I've read your helpful readme a few times, but it'd be nice to understand what each of the classes is providing. Thank you.

**Comment by waacton** [comment]
There's no additional documentation at the moment, but I'll continue to think about how I can improve it!

I've avoided something like XML documentation until now because I expect I would get bogged down in trying to explain intricacies of colour theory in text snippets around the codebase, whereas the readme can be more demonstration-based.

I've tried to keep the API itself fairly simple and accessible - and hopefully quite natural to explore in code - but it's grown slowly over time, and I can see the benefit of having a reference to all available properties and functions at a glance. I'll keep working at it :+1:
